### PR TITLE
V4 build update

### PR DESCRIPTION
**Pull request by baallan**
This fixes missing dependencies that break the build on ubuntu 18.

### CODE REVIEW

File: ldms/src/ldmsd/Makefile.am [hunk 1]
```diff
@@ -22,12 +22,14 @@ AM_CFLAGS += -DPLUGINDIR='"$(pkglibdir)"'
 
 ldms_ls_SOURCES = ldms_ls.c
 ldms_ls_LDADD = -lldms -lzap -lmmalloc -lovis_util -lcoll \
-        -lpthread \
+        -ljson_util -lovis_third -lpthread \
```
Comment (narategithub) [inline ldms/src/ldmsd/Makefile.am]: Why adding ovis_third to ldms_ls? I don't think it uses the library.

File: ldms/src/ldmsd/Makefile.am [hunk 2]
```diff
@@ -22,12 +22,14 @@ AM_CFLAGS += -DPLUGINDIR='"$(pkglibdir)"'
...
         @LDFLAGS_GETTIME@
 
 lib_LTLIBRARIES += libldmsd_stream.la
 libldmsd_stream_la_SOURCES = ldmsd_stream.c ldmsd_stream.h ldmsd_request_util.c
-libldmsd_stream_la_LIBADD = -lldms
+libldmsd_stream_la_LIBADD = -lldms \
+	-lzap -lmmalloc -lovis_util -lcoll -ljson_util \
+	-lovis_event -lovis_ctrl -lovis_third -lm -ldl -lpthread
```
Comment (narategithub) [inline ldms/src/ldmsd/Makefile.am]: ovis_third?

File: ldms/src/ldmsd/Makefile.am [hunk 3]
```diff
@@ -40,7 +42,7 @@ ldmsd_SOURCES = ldmsd.c ldmsd_config.c \
 	ldmsd_failover.c ldmsd_group.c ldmsd_auth.c
 ldmsd_LDADD = -lldms libldmsd_request.la libldmsd_stream.la \
 	-lzap -lmmalloc -lovis_util -lcoll -ljson_util \
-	-lovis_event -lpthread -lovis_ctrl -lm -ldl
+	-lovis_event -lovis_ctrl -lovis_third -lm -ldl -lpthread
```
Comment (narategithub) [inline ldms/src/ldmsd/Makefile.am]: ovis_third?

File: ldms/src/ldmsd/Makefile.am [hunk 4]
```diff
@@ -51,8 +53,8 @@ endif
 
 ldmsctl_SOURCES = ldmsctl.c
 ldmsctl_LDADD = -lldms libldmsd_request.la \
-		-lovis_ctrl -lovis_util -lmmalloc -lcoll -lzap -ljson_util \
-		-lm -lpthread @LDFLAGS_GETTIME@ @LDFLAGS_READLINE@
+		-lovis_ctrl -lovis_util -lmmalloc -lzap -ljson_util -lcoll \
+		-lovis_third -lm -lpthread @LDFLAGS_GETTIME@ @LDFLAGS_READLINE@
```
Comment (narategithub) [inline ldms/src/ldmsd/Makefile.am]: ovis_third?

File: lib/src/json/Makefile.am [hunk 5]
```diff
@@ -32,6 +32,6 @@ lib_LTLIBRARIES += libjson_util.la
 
 json_test_SOURCES = json_test.c json_util.h
 json_test_CFLAGS = $(AM_CFLAGS) -g -O0
-json_test_LDADD = libjson_util.la
+json_test_LDADD = libjson_util.la -lcoll -lovis_util -lovis_third -lc -lcrypto
```
Comment (narategithub) [inline lib/src/json/Makefile.am]: ovis_third? crypto?

File: ldms/src/ldmsd/test/Makefile.am [hunk 6]
```diff
@@ -6,9 +6,9 @@ AM_CPPFLAGS = @OVIS_INCLUDE_ABS@
 
 sbin_PROGRAMS += ldmsd_stream_publish
 ldmsd_stream_publish_SOURCES = ldmsd_stream_publish.c
-ldmsd_stream_publish_LDADD = -lldms -lldmsd_request -lldmsd_stream -ljson_util -lovis_util
+ldmsd_stream_publish_LDADD = -lldms -lldmsd_request -lldmsd_stream -ljson_util -lovis_util -lovis_third -lcoll -lzap -lmmalloc
```
Comment (narategithub) [inline ldms/src/ldmsd/test/Makefile.am]: ovis_third?

File: ldms/src/ldmsd/test/Makefile.am [hunk 7]
```diff
@@ -6,9 +6,9 @@ AM_CPPFLAGS = @OVIS_INCLUDE_ABS@
...
 
 sbin_PROGRAMS += ldmsd_stream_subscribe
 ldmsd_stream_subscribe_SOURCES = ldmsd_stream_subscribe.c
-ldmsd_stream_subscribe_LDADD = -lldms -lldmsd_request -lldmsd_stream -ljson_util -lovis_util
+ldmsd_stream_subscribe_LDADD = -lldms -lldmsd_request -lldmsd_stream -ljson_util -lovis_util -lovis_third -lcoll -lzap -lmmalloc
```
Comment (narategithub) [inline ldms/src/ldmsd/test/Makefile.am]: ovis_third?

**Comment by baallan** [comment]
Ubuntu 18.04 build fails without the additions (tool versions below).
coll depends on third.

```
gcc --version 
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

automake --version
automake (GNU automake) 1.15.1

libtool --version
libtool (GNU libtool) 2.4.6
```

**Comment by narategithub** [comment]
Hmm... I don't get any error when building OVIS-4 with the following configure flags:
```
./configure --prefix=/opt/ovis --enable-etc
```

Some info on my build:

```
root@ubuntu-1804:~/ovis/build# git show --stat                                 
commit feb00cca95fde89ac36b8890f825a097a99787ce (HEAD -> OVIS-4, origin/OVIS-4)

root@ubuntu-1804:~/ovis/build# gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

root@ubuntu-1804:~/ovis/build# automake --version
automake (GNU automake) 1.15.1

root@ubuntu-1804:~/ovis/build# libtoolize --version
libtoolize (GNU libtool) 2.4.6

```

**Comment by narategithub** [comment]
Adding `CFLAGS='-Wall -Werror'` also works.

**Comment by baallan** [comment]
```
../configure --prefix=/home/baallan/ovis/repos/ovis-papi/LDMS_install --disable-rdma --enable-ssl --disable-sos --disable-perfevent --disable-rpath --enable-zap --enable-zaptest --disable-swig --enabl
e-ovis_auth --enable-sysclassib --enable-syspapi-sampler
```

gets

```
make[6]: Entering directory '/home/baallan/ovis/repos/ovis-papi/LDMS_objdir/lib/src/json'
  CC       json_util.lo
flex -o json_lexer.c ../../../../lib/src/json/json_lexer.l
  CC       json_lexer.lo
  CC       json_parser.lo
../../../../lib/src/json/json_parser.y: In function ‘json_parser_new’:
../../../../lib/src/json/json_parser.y:133:3: warning: implicit declaration of function ‘yylex_init’; did you mean ‘yylex’? [-Wimplicit-function-declaration]
   yylex_init(&p->scanner);
   ^~~~~~~~~~
   yylex
../../../../lib/src/json/json_parser.y: In function ‘json_parser_free’:
../../../../lib/src/json/json_parser.y:139:2: warning: implicit declaration of function ‘yylex_destroy’ [-Wimplicit-function-declaration]
  yylex_destroy(parser->scanner);
  ^~~~~~~~~~~~~
  CCLD     libjson_util.la
ar: `u' modifier ignored since `D' is the default (see `U')
  CC       json_test-json_test.o
  CCLD     json_test
/usr/bin/ld: warning: libcoll.so.0, needed by ./.libs/libjson_util.so, not found (try using -rpath or -rpath-link)
./.libs/libjson_util.so: undefined reference to `htbl_next'
./.libs/libjson_util.so: undefined reference to `htbl_alloc'
./.libs/libjson_util.so: undefined reference to `htbl_free'
./.libs/libjson_util.so: undefined reference to `htbl_first'
./.libs/libjson_util.so: undefined reference to `htbl_find'
./.libs/libjson_util.so: undefined reference to `hent_init'
./.libs/libjson_util.so: undefined reference to `htbl_empty'
./.libs/libjson_util.so: undefined reference to `htbl_ins'
./.libs/libjson_util.so: undefined reference to `htbl_del'
collect2: error: ld returned 1 exit status
```
```

**Comment by baallan** [comment]
most likely the relevant config flag to test with is --disable-rpath

**Comment by tom95858** [comment]
libcoll should not depend on libthird. This kind of thing requires every user of a given library to know every dependency of all of the libraries that it uses. This is impossible to maintain. 

Please be specific about what in libcoll requires libthird. It should be nothing.

**Comment by baallan** [comment]
libcoll uses libthird for third party string hashing functions in label-set.c (and has since v2).
Except for license documentation reasons (which is a bit of a flimsy reason), libthird could be folded into libcoll.

There's nothing 'impossible to maintain' about the current layout, since the proposed patch maintains it. 

A much better resolution to the issue (as @morrone proposed at some point, I believe) is that we merge together most of the base libraries of ldmsd into one, instead of maintaining this micro-fragmented lib/ project which is only used by ldms. (e.g. sos maintains its own data structure libraries).
I'd be happy to put the defragmentation project on my todo list.

**Comment by narategithub** [comment]
I think you did not address the root cause in this patch. Please see if #205 fix the problem.

**Comment by tom95858** [comment]
I would much prefer if we either bound libcoll with libthird non-shared, or as @baallan suggests implement the one function (b62_encode afaict) that it uses  in the label-set.c file and blow away libthird altogether.

That said, as I mentioned, in general, it is bad form for a user of libcoll, e.g. libjson_util to have to know what libcoll depends on and add it to it's makefile.

**Comment by baallan** [comment]
The bigger wl,rpath change by Narate has solved the same problem a better way.

**Pull request closed without merging**